drivers/azure/azure.go

```diff
@@ -0,0 +1,501 @@
+	if exists, err := c.VirtualMachineExists(d.ResourceGroup, d.naming().VM()); err != nil {
+		return err
+	} else if exists {
+		return fmt.Errorf("Virtual Machine with name %s already exists in resource group %q", d.naming().VM(), d.ResourceGroup)
+	}
+
+	// NOTE(ahmetalpbalkan) we could have done more checks here but Azure often
+	// returns meaningful error messages and it would be repeating the backend
+	// logic on the client side. Some examples:
+	//   - Deployment of a machine to an existing Virtual Network fails if
+	//     virtual network is in a different region.
+	//   - Changing IP Address space of a subnet would fail if there are machines
+	//     running in the Virtual Network.
+	log.Info("Completed machine pre-create checks.")
+	return nil
+}
+
+// Create creates the virtual machine.
+func (d *Driver) Create() error {
+	// NOTE(ahmetalpbalkan): We can probably parallelize the sh*t out of this.
+	// However that would lead to a concurrency logic and while creation of a
+	// resource fails, other ones would be kicked off, which could lead to a
+	// resource leak. This is slower but safer.
+	c, err := d.newAzureClient()
+	if err != nil {
+		return err
+	}
+
+	if err := c.CreateResourceGroup(d.ResourceGroup, d.Location); err != nil {
+		return err
+	}
+	if err := c.CreateAvailabilitySetIfNotExists(d.ctx, d.ResourceGroup, d.AvailabilitySet, d.Location); err != nil {
+		return err
+	}
+	if err := c.CreateNetworkSecurityGroup(d.ctx, d.ResourceGroup, d.naming().NSG(), d.Location, d.ctx.FirewallRules); err != nil {
+		return err
+	}
+	if err := c.CreateVirtualNetworkIfNotExists(d.ResourceGroup, d.VirtualNetwork, d.Location); err != nil {
+		return err
+	}
+	if err := c.CreateSubnet(d.ctx, d.ResourceGroup, d.VirtualNetwork, d.SubnetName, d.SubnetPrefix); err != nil {
+		return err
+	}
+	if d.NoPublicIP {
+		log.Info("Not creating a public IP address.")
+	} else {
+		if err := c.CreatePublicIPAddress(d.ctx, d.ResourceGroup, d.naming().IP(), d.Location); err != nil {
+			return err
+		}
+	}
+	if err := c.CreateNetworkInterface(d.ctx, d.ResourceGroup, d.naming().NIC(), d.Location,
+		d.ctx.PublicIPAddressID, d.ctx.SubnetID, d.ctx.NetworkSecurityGroupID, d.PrivateIPAddr); err != nil {
+		return err
+	}
+	if err := c.CreateStorageAccount(d.ctx, d.ResourceGroup, d.Location, defaultStorageType); err != nil {
+		return err
+	}
+	if err := d.generateSSHKey(d.ctx); err != nil {
+		return err
+	}
+	if err := c.CreateVirtualMachine(d.ResourceGroup, d.naming().VM(), d.Location, d.Size, d.ctx.AvailabilitySetID,
+		d.ctx.NetworkInterfaceID, d.BaseDriver.SSHUser, d.ctx.SSHPublicKey, d.Image, d.ctx.StorageAccount); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Remove deletes the virtual machine and resources associated to it.
+func (d *Driver) Remove() error {
+	if err := d.checkLegacyDriver(false); err != nil {
+		return err
+	}
+
+	// NOTE(ahmetalpbalkan):
+	//   - remove attemps are best effort and if a resource is already gone, we
```

Comment:
I meant I guess there's a typo. The question marks is there because I'm not sure of my english ;-)